File: python/aethernet/identity/erid_announcement_codec.py
# SPDX-License-Identifier: MIT

"""Frames the in-session ERID announcement — the message a node sends a peer INSIDE an
established Signal session to share its secret routing key (plus the rotation parameters
it uses), so the peer can resolve its rotating wire address via :class:`EridDirectory`.

The bytes are carried *encrypted* by the Signal session, so this is framing only — no
encryption of its own. A 4-byte magic sentinel + version lets a receiver tell an ERID
announcement apart from other in-session application data before trying to parse it.

Layout: magic ``AERD`` (4) + version (1) + ``epoch_seconds`` (int32 BE) +
``erid_length`` (int32 BE) + ``routing_key_len`` (int32 BE) + ``routing_key``. Integer
fields are big-endian so every language port frames byte-identically. Port of the C#
reference (``src/AetherNet.Core/Identity/EridAnnouncementCodec.cs``).
"""

from __future__ import annotations

import struct
from typing import NamedTuple

from aethernet.identity.ephemeral_routing_id import DEFAULT_EPOCH_SECONDS, DEFAULT_LENGTH

_MAGIC = b"AERD"  # 0x41 0x45 0x52 0x44
_VERSION = 1
# magic(4) + version(1) + epoch_seconds(4) + erid_length(4) + routing_key_len(4) = 17.
_HEADER_LENGTH = 17


class EridAnnouncement(NamedTuple):
    """A decoded in-session ERID announcement."""

    routing_key: bytes
    epoch_seconds: int
    erid_length: int


def encode(
    routing_key: bytes,
    epoch_seconds: int = DEFAULT_EPOCH_SECONDS,
    erid_length: int = DEFAULT_LENGTH,
) -> bytes:
    """Frame an announcement carrying ``routing_key`` and the rotation params.

    Raises
    ------
    ValueError
        If ``routing_key`` is empty, ``epoch_seconds`` is not positive, or
        ``erid_length`` is outside ``1..51``.
    """
    if not routing_key:
        raise ValueError("routing_key cannot be empty")
    if epoch_seconds <= 0:
        raise ValueError("epoch_seconds must be positive")
    if erid_length < 1 or erid_length > 51:
        raise ValueError("erid_length must be 1..51")

    # '>' = big-endian; B = version (uint8); three signed int32 fields, matching the C#
    # BinaryPrimitives.WriteInt32BigEndian calls byte-for-byte.
    header = _MAGIC + struct.pack(
        ">Biii", _VERSION, epoch_seconds, erid_length, len(routing_key)
    )
    return header + bytes(routing_key)


def try_decode(data: bytes) -> EridAnnouncement | None:
    """Parse an announcement.

    Returns ``None`` (rather than raising) when the bytes are not a well-formed ERID
    announcement, so a receiver can cheaply test an arbitrary decrypted in-session
    payload against the magic.
    """
    if len(data) < _HEADER_LENGTH:
        return None
    if data[0:4] != _MAGIC:
        return None
    if data[4] != _VERSION:
        return None

    epoch_seconds, erid_length, key_len = struct.unpack(">iii", data[5:17])

    if epoch_seconds <= 0:
        return None
    if erid_length < 1 or erid_length > 51:
        return None
    if key_len <= 0 or _HEADER_LENGTH + key_len > len(data):
        return None

    return EridAnnouncement(
        routing_key=bytes(data[_HEADER_LENGTH : _HEADER_LENGTH + key_len]),
        epoch_seconds=epoch_seconds,
        erid_length=erid_length,
    )

File: kotlin/src/main/kotlin/aethernet/identity/EridAnnouncementCodec.kt
// SPDX-License-Identifier: MIT

package aethernet.identity

/**
 * Frames the in-session ERID announcement — the message a node sends a peer INSIDE an
 * established Signal session to share its secret `routingKey` (plus the rotation parameters it
 * uses), so the peer can resolve its rotating wire address via [EridDirectory].
 *
 * The bytes are carried *encrypted* by the Signal session, so this is framing only — no
 * encryption of its own. A 4-byte magic sentinel + version lets a receiver tell an ERID
 * announcement apart from other in-session application data before trying to parse it.
 *
 * Layout: magic `AERD` (4) + version (1) + epochSeconds (Int32 BE) + eridLength (Int32 BE) +
 * routingKeyLen (Int32 BE) + routingKey. Integer fields big-endian so every language port
 * frames byte-identically. Port of the C# reference.
 */
object EridAnnouncementCodec {

    /** A decoded in-session ERID announcement. */
    class Announcement(
        val routingKey: ByteArray,
        val epochSeconds: Int,
        val eridLength: Int,
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is Announcement) return false
            return routingKey.contentEquals(other.routingKey) &&
                epochSeconds == other.epochSeconds &&
                eridLength == other.eridLength
        }

        override fun hashCode(): Int =
            (routingKey.contentHashCode() * 31 + epochSeconds) * 31 + eridLength
    }

    // 'A' 'E' 'R' 'D' — "AetherNet ERID Directory announcement".
    private val MAGIC = byteArrayOf(0x41, 0x45, 0x52, 0x44)
    private const val VERSION: Byte = 1
    // magic(4) + version(1) + epochSeconds(4) + eridLength(4) + routingKeyLen(4).
    private const val HEADER_LENGTH = 17

    /**
     * Frame an announcement carrying [routingKey] and the rotation params.
     *
     * @throws IllegalArgumentException if [routingKey] is empty, [epochSeconds] is not
     *   positive, or [eridLength] is outside 1..51.
     */
    fun encode(
        routingKey: ByteArray,
        epochSeconds: Int = EphemeralRoutingId.DEFAULT_EPOCH_SECONDS.toInt(),
        eridLength: Int = EphemeralRoutingId.DEFAULT_LENGTH,
    ): ByteArray {
        require(routingKey.isNotEmpty()) { "routingKey cannot be empty" }
        require(epochSeconds > 0) { "epochSeconds must be positive" }
        require(eridLength in 1..51) { "eridLength must be 1..51" }

        val buf = ByteArray(HEADER_LENGTH + routingKey.size)
        System.arraycopy(MAGIC, 0, buf, 0, 4)
        buf[4] = VERSION
        writeInt32BE(buf, 5, epochSeconds)
        writeInt32BE(buf, 9, eridLength)
        writeInt32BE(buf, 13, routingKey.size)
        System.arraycopy(routingKey, 0, buf, HEADER_LENGTH, routingKey.size)
        return buf
    }

    /**
     * Parse an announcement. Returns null (rather than throwing) when the bytes are not a
     * well-formed ERID announcement, so a receiver can cheaply test an arbitrary decrypted
     * in-session payload against the magic.
     */
    fun tryDecode(data: ByteArray): Announcement? {
        if (data.size < HEADER_LENGTH) return null
        for (i in 0 until 4) if (data[i] != MAGIC[i]) return null
        if (data[4] != VERSION) return null

        val epochSeconds = readInt32BE(data, 5)
        val eridLength = readInt32BE(data, 9)
        val keyLen = readInt32BE(data, 13)

        if (epochSeconds <= 0) return null
        if (eridLength < 1 || eridLength > 51) return null
        if (keyLen <= 0 || HEADER_LENGTH.toLong() + keyLen.toLong() > data.size.toLong()) return null

        return Announcement(data.copyOfRange(HEADER_LENGTH, HEADER_LENGTH + keyLen), epochSeconds, eridLength)
    }

    private fun writeInt32BE(buf: ByteArray, offset: Int, value: Int) {
        buf[offset] = ((value ushr 24) and 0xFF).toByte()
        buf[offset + 1] = ((value ushr 16) and 0xFF).toByte()
        buf[offset + 2] = ((value ushr 8) and 0xFF).toByte()
        buf[offset + 3] = (value and 0xFF).toByte()
    }

    private fun readInt32BE(buf: ByteArray, offset: Int): Int =
        ((buf[offset].toInt() and 0xFF) shl 24) or
            ((buf[offset + 1].toInt() and 0xFF) shl 16) or
            ((buf[offset + 2].toInt() and 0xFF) shl 8) or
            (buf[offset + 3].toInt() and 0xFF)
}
